src/cowboy_decompress_h.erl

```diff
@@ -0,0 +1,217 @@
+	fold(Commands, State#state{next=Next}).
+
+-spec terminate(cowboy_stream:streamid(), cowboy_stream:reason(), #state{}) -> any().
+terminate(StreamID, Reason, #state{next=Next, inflate=Z}) ->
+	case Z of
+		undefined -> ok;
+		_ -> zlib:close(Z)
+	end,
+	cowboy_stream:terminate(StreamID, Reason, Next).
+
+-spec early_error(cowboy_stream:streamid(), cowboy_stream:reason(),
+	cowboy_stream:partial_req(), Resp, cowboy:opts()) -> Resp
+	when Resp::cowboy_stream:resp_command().
+early_error(StreamID, Reason, PartialReq, Resp, Opts) ->
+	cowboy_stream:early_error(StreamID, Reason, PartialReq, Resp, Opts).
+
+%% Internal.
+
+check_req(Req) ->
+	try cowboy_req:parse_header(<<"content-encoding">>, Req) of
+		undefined ->
+			#state{compress=undefined};
+		Encodings ->
+			case [E || E=(<<"gzip">>) <- Encodings] of
+				[] ->
+					#state{compress=undefined};
+				_ ->
+					#state{compress=gzip}
+			end
+	catch
+		_:_ ->
+			#state{compress=undefined}
+	end.
+
+fold(Commands, State) ->
+	fold(Commands, State, []).
+
+fold([], State, Acc) ->
+	{lists:reverse(Acc), State};
+fold([{response, Status, Headers0, Body}|Tail], State=#state{ignore=false}, Acc) ->
+	Headers = add_accept_encoding(Headers0),
+	fold(Tail, State, [{response, Status, Headers, Body}|Acc]);
+fold([{headers, Status, Headers0} | Tail], State=#state{ignore=false}, Acc) ->
+	Headers = add_accept_encoding(Headers0),
+	fold(Tail, State, [{headers, Status, Headers}|Acc]);
+fold([Command|Tail], State, Acc) ->
+	fold(Tail, State, [Command|Acc]).
+
+add_accept_encoding(Headers=#{<<"accept-encoding">> := AcceptEncoding}) ->
+	try cow_http_hd:parse_accept_encoding(iolist_to_binary(AcceptEncoding)) of
+		List ->
+			case lists:keyfind(<<"gzip">>, 1, List) of
+				%% gzip is excluded but this handler is not ignored; we replace.
+				{_, 0} ->
+					Replaced = lists:keyreplace(<<"gzip">>, 1, List, {<<"gzip">>, 1000}),
+					Codings = build_accept_encoding(Replaced),
+					Headers#{<<"accept-encoding">> => Codings};
+				{_, _} ->
+					Headers;
+				false ->
+					case lists:keyfind(<<"*">>, 1, List) of
+						%% Others are excluded along with gzip; we add.
+						{_, 0} ->
+							WithGzip = [{<<"gzip">>, 1000} | List],
+							Codings = build_accept_encoding(WithGzip),
+							Headers#{<<"accept-encoding">> => Codings};
+						{_, _} ->
+							Headers;
+						false ->
+							Headers#{<<"accept-encoding">> => [AcceptEncoding, <<", gzip">>]}
+					end
+			end
+	catch _:_ ->
+		Headers#{<<"accept-encoding">> => <<"gzip">>}
+	end;
+add_accept_encoding(Headers) ->
+	Headers#{<<"accept-encoding">> => <<"gzip">>}.
+
+%% From cowlib, maybe expose?
+qvalue_to_iodata(0) -> <<"0">>;
+qvalue_to_iodata(Q) when Q < 10 -> [<<"0.00">>, integer_to_binary(Q)];
+qvalue_to_iodata(Q) when Q < 100 -> [<<"0.0">>, integer_to_binary(Q)];
+qvalue_to_iodata(Q) when Q < 1000 -> [<<"0.">>, integer_to_binary(Q)];
+qvalue_to_iodata(1000) -> <<"1">>.
+
+build_accept_encoding([{ContentCoding, Q}|Tail]) ->
```

Comment:
I'll add to Cowlib instead.

Reply:
I'll leave a todo and add to Cowlib in a future release. I'll avoid doing Cowlib changes this time around.